### DISCUSSION TITLE
First music

### DISCUSSION THREAD

**Discussion by jkominek** [jkominek/piano-conversion#49] [Show and tell]
Playing a quick tune on 11 keys, with 3 octaves of shift.

https://youtu.be/GiBngxTWHDI

I'll post it to piano clack later, but you dedicated followers get to see it first. :)

**Comment by davidedelvento** [comment]
Awesome! Working on something similar myself, today!

**Comment by jkominek** [comment]
Picking up from https://pianoclack.com/forum/d/289-scanning-speed-and-velocity-mapping-of-cybrid/75 and elaborating slightly.

The firmware for computing the pseudovelocities starts [here](https://github.com/jkominek/piano-conversion/blob/a8b33c9164264ce179180f5ef9f337700e9bbd1d/firmware/adc-board/Core/Src/main.c#L1047).

The PC side of things is [here](https://github.com/jkominek/piano-conversion/blob/master/firmware/adc-board/adcreceive.py).

It occurred to me that my scaling in the Python maxes out at generating 7 "m/s" values to feed into the Goebl-Bresin equation. But 127 corresponds to 9.29 m/s. I wouldn't have expected to see anything past 118. (I didn't read the graph quite right when I was grinding out that bit of code.)

So I'll need to play around with the scaling range, and hopefully wire up some more keys soon so I can try playing some real two handed music to see if things feel reasonable to me. I can dragoon in a friend who is a significantly more experienced pianist to give it a shot at that point, too.

And for reference, the VST in use in the demo was Pianoteq 7.4.1 with the "C. Bechstein DG Prelude" configuration. No further tweaking has been performed; the velocity curve is perfectly flat. The MIDI configuration panel confirms that it was interpreting the CC#88 commands, as the on/off messages all include fractional velocities.

## Next steps

Now that the firmware / algorithm actually seem mostly reasonable, I need to finish putting in place the functionality the board will need for automated calibration, and then write the PC calibrator software. That'll get more reasonable cutoff and scaling values, which should make the MIDI output even better.

In parallel with that I expect I'll start assembling more boards / cables. I'm about to order my remaining CNY70s, and I think I have all the other necessary parts in stock.
